### PR TITLE
Add SSL Pinning support

### PR DESCRIPTION
**Pull request by colinmcardell**
For reference: https://github.com/spotify/SPTDataLoader/issues/109

### CODE REVIEW

**Comment by 8W9aG** [comment]
:+1: Looks really great, needs a couple of unit tests though. One thing I have been thinking about is restricting the certificate lookups to using NSBundle pathForResource, it might be advantageous to only allow certificates from a read-only directory on the device (however a minor point)

File: SPTDataLoader/SPTDataLoaderServerTrustPolicy.m [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        SecCertificateRef certificate = SecTrustGetCertificateAtIndex(trust, i);
+        [certificates addObject:(__bridge_transfer NSData *)SecCertificateCopyData(certificate)];
+    }
+    
+    return [NSArray arrayWithArray:certificates];
+}
+
+@interface SPTDataLoaderServerTrustPolicy ()
+
+@property  (strong, nonatomic) NSDictionary<NSString *, NSArray<NSData *> *> *trustedHostsAndCertificates;
```
Comment (8W9aG) [inline SPTDataLoader/SPTDataLoaderServerTrustPolicy.m]: Might be good to switch around nonatomic/strong so it conforms to the style of the rest of the repository

File: SPTDataLoader/SPTDataLoaderServerTrustPolicy.m [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    
+    return [certificates copy];
+}
+
+#pragma mark Lifecycle
+
+- (instancetype)initWithHostsAndCertificatePaths:(NSDictionary<NSString *, NSArray<NSString *> *> *)hostsAndCertificatePaths
+{
+    self = [super init];
+    if (!self) {
```
Comment (8W9aG) [inline SPTDataLoader/SPTDataLoaderServerTrustPolicy.m]: We generally do if (self) { // initialisation } (see the ios-style repo)

File: include/SPTDataLoader/SPTDataLoaderServerTrustPolicy.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @param hostsAndCertificatePaths An NSDictionary where the key is the host and
+ *                                 the value is an NSArray of paths to
+ *                                 certificate files that will be used while
+ *                                 validating a challenge.
+ * @discussion The NSString key for host can include wildcards
+ */
++ (instancetype)policyWithHostsAndCertificatePaths:(NSDictionary<NSString *, NSArray<NSString *> *> *)hostsAndCertificatePaths;
+
+- (BOOL)validateChallenge:(NSURLAuthenticationChallenge *)challenge;
```
Comment (8W9aG) [inline include/SPTDataLoader/SPTDataLoaderServerTrustPolicy.h]: Remember to add a small doc string for this function

**Comment by 8W9aG** [comment]
👍 This is looking great, i think the only thing it needs is a rebase and then we can merge it

**Comment by colinmcardell** [comment]
@8W9aG Awesome!
Did you want me to squash and force push down to a single commit??

**Comment by 8W9aG** [comment]
Nah I think we should leave the history as is, if you are fine with this PR I'll merge it

**Comment by colinmcardell** [comment]
I'm into it. Merge away! Thanks!

**Pull request merged**